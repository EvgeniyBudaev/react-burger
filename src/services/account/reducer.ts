import { AxiosError } from "axios";
import { Reducer } from "redux";
import { AccountActionsType, ActionTypes } from "services/account";
import { IUser } from "./types";

interface IAccountState {
    accessToken: string | null;
    refreshToken: string | null;
    emailSent: boolean;
    passwordReseted: boolean;
    success: boolean;
    user: IUser | null;
    forgotPasswordRequest: boolean;
    forgotPasswordFailed: boolean;
    loginRequest: boolean;
    loginFailed: boolean;
    registerRequest: boolean;
    registerFailed: boolean;
    resetPasswordRequest: boolean;
    resetPasswordFailed: boolean;
    error: AxiosError | null;
}

const initialState: IAccountState = {
    accessToken: null,
    refreshToken: null,
    emailSent: false,
    passwordReseted: false,
    success: false,
    user: null,
    forgotPasswordRequest: false,
    forgotPasswordFailed: false,
    loginRequest: false,
    loginFailed: false,
    registerRequest: false,
    registerFailed: false,
    resetPasswordRequest: false,
    resetPasswordFailed: false,
    error: null,
};

export const reducer: Reducer<IAccountState, AccountActionsType> = (
    state = initialState,
    action
) => {
    switch (action.type) {
        case ActionTypes.LOGIN_USER_REQUEST: {
            return {
                ...state,
                loginRequest: true,
                loginFailed: false,
                error: null,
                success: false,
            };
        }
        case ActionTypes.LOGIN_USER_SUCCESS:
            return {
                ...state,
                accessToken: action.payload.accessToken,
                refreshToken: action.payload.refreshToken,
                loginRequest: false,
                success: action.payload.success,
                user: action.payload.user,
            };
        case ActionTypes.LOGIN_USER_FAILED:
            return {
                ...state,
                loginRequest: false,
                loginFailed: true,
                error: action.payload,
            };
        case ActionTypes.REGISTER_USER_REQUEST: {
            return {
                ...state,
                registerRequest: true,
                registerFailed: false,
                error: null,
                success: false,
            };
        }
        case ActionTypes.REGISTER_USER_SUCCESS:
            return {
                ...state,
                accessToken: action.payload.accessToken,
                refreshToken: action.payload.refreshToken,
                registerRequest: false,
                success: action.payload.success,
                user: action.payload.user,
            };
        case ActionTypes.REGISTER_USER_FAILED:
            return {
                ...state,
                registerRequest: false,
                registerFailed: true,
                error: action.payload,
            };
        case ActionTypes.FORGOT_PASSWORD_REQUEST: {
            return {
                ...state,
                emailSent: false,
                forgotPasswordRequest: true,
                forgotPasswordFailed: false,
                error: null,
            };
        }
        case ActionTypes.FORGOT_PASSWORD_SUCCESS:
            return {
                ...state,
                emailSent: true,
                forgotPasswordRequest: false,
            };
        case ActionTypes.FORGOT_PASSWORD_FAILED:
            return {
                ...state,
                forgotPasswordRequest: false,
                forgotPasswordFailed: true,
                error: action.payload,
            };
        case ActionTypes.RESET_PASSWORD_REQUEST: {
            return {
                ...state,
                passwordReseted: false,
                resetPasswordRequest: true,
                resetPasswordFailed: false,
                error: null,
            };
        }
        case ActionTypes.RESET_PASSWORD_SUCCESS:
            return {
                ...state,
                passwordReseted: true,
                resetPasswordRequest: false,
            };
        case ActionTypes.RESET_PASSWORD_FAILED:
            return {
                ...state,
                resetPasswordRequest: false,
                resetPasswordFailed: true,
                error: action.payload,
            };
        default:
            return state;
    }
};
