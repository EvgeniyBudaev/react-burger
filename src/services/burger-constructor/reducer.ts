import { INGREDIENT_TYPE } from "constants/ingredient";
import { Reducer } from "redux";
import update from "immutability-helper";
import { IIngredient } from "types/ingredient";
import { ActionTypes } from "./actionTypes";
import { BurgerConstructorActionsType } from "./types";

interface IBurgerConstructorState {
    bun: IIngredient | Record<string, never>;
    mains: IIngredient[];
}

const initialState: IBurgerConstructorState = {
    bun: {},
    mains: [],
};

export const reducer: Reducer<
    IBurgerConstructorState,
    BurgerConstructorActionsType
> = (state = initialState, action) => {
    switch (action.type) {
        case ActionTypes.ADD_INGREDIENT: {
            if (action.payload.type === INGREDIENT_TYPE.BUN) {
                return {
                    ...state,
                    bun: action.payload,
                };
            } else
                return {
                    ...state,
                    mains: [...state.mains, action.payload],
                };
        }
        case ActionTypes.DELETE_INGREDIENT: {
            return {
                ...state,
                mains: update(state.mains, {
                    $splice: [[action.payload, 1]],
                }),
            };
        }
        case ActionTypes.MOVE_INGREDIENTS: {
            const draggingIngredient = state.mains[action.dragIndex];
            return {
                ...state,
                mains: update(state.mains, {
                    $splice: [
                        [action.dragIndex, 1],
                        [action.hoverIndex, 0, draggingIngredient],
                    ],
                }),
            };
        }
        default:
            return state;
    }
};
