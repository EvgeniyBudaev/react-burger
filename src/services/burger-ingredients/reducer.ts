import { Reducer } from "redux";
import { IError } from "types/error";
import { IIngredient } from "types/ingredient";
import {
    ActionTypes,
    BurgerIngredientsActionsType,
} from "services/burger-ingredients/index";

export interface IBurgerIngredientsState {
    ingredients: IIngredient[];
    ingredientsRequest: boolean;
    ingredientsFailed: boolean;
    ingredientsError: IError | null;
}

const initialState: IBurgerIngredientsState = {
    ingredients: [],
    ingredientsRequest: false,
    ingredientsFailed: false,
    ingredientsError: null,
};

export const reducer: Reducer<
    IBurgerIngredientsState,
    BurgerIngredientsActionsType
> = (state = initialState, action) => {
    switch (action.type) {
        case ActionTypes.GET_BURGER_INGREDIENTS_REQUEST: {
            return {
                ...state,
                ingredientsRequest: true,
                ingredientsFailed: false,
                ingredientsError: null,
            };
        }
        case ActionTypes.GET_BURGER_INGREDIENTS_SUCCESS:
            return {
                ...state,
                ingredients: [...action.payload].map(item => {
                    return { ...item, count: 0 };
                }),
                ingredientsRequest: false,
            };
        case ActionTypes.GET_BURGER_INGREDIENTS_FAILED:
            return {
                ...state,
                ingredientsRequest: false,
                ingredientsFailed: true,
                ingredientsError: action.payload,
            };
        default:
            return state;
    }
};
