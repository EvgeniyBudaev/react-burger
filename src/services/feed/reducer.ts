import { Reducer } from "redux";
import { ActionTypes, TFeedActions } from "services/feed";
import { TFeed } from "types/feed";

interface IFeedState {
    wsFeedRequest: boolean;
    wsFeedConnected: boolean;
    feed: TFeed | null;
    feedRequest: boolean;
    feedFailed: boolean;
    error: Event | null;
}

const initialState: IFeedState = {
    wsFeedRequest: false,
    wsFeedConnected: false,
    feed: null,
    feedRequest: false,
    feedFailed: false,
    error: null,
};

export const reducer: Reducer<IFeedState, TFeedActions> = (
    state = initialState,
    action
) => {
    switch (action.type) {
        case ActionTypes.WS_FEED_CONNECTION_START: {
            return {
                ...state,
                wsFeedRequest: true,
                error: null,
            };
        }
        case ActionTypes.WS_FEED_CONNECTION_STOP: {
            return {
                ...state,
                wsFeedConnected: false,
            };
        }
        case ActionTypes.WS_FEED_CONNECTION_SUCCESS: {
            return {
                ...state,
                wsFeedConnected: true,
                error: null,
            };
        }
        case ActionTypes.WS_FEED_CONNECTION_ERROR: {
            return {
                ...state,
                wsFeedRequest: false,
                error: action.payload,
            };
        }
        case ActionTypes.WS_FEED_CONNECTION_CLOSED: {
            return {
                ...state,
                wsFeedConnected: false,
                error: null,
            };
        }
        case ActionTypes.WS_GET_FEED: {
            return {
                ...state,
                wsFeedRequest: false,
                feed: {
                    ...state.feed,
                    orders: action.payload.orders,
                    success: true,
                    total: action.payload.total,
                    totalToday: action.payload.totalToday,
                },
            };
        }
        default:
            return state;
    }
};
