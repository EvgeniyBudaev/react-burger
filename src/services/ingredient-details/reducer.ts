import { Reducer } from "redux";
import { ActionTypes } from "./actionTypes";
import { IngredientDetailsActionsType } from "./types";

interface IIngredientDetailsState {
    ingredientDetailsId: string;
    iSIngredientDetailsActive: boolean;
}

const initialState: IIngredientDetailsState = {
    ingredientDetailsId: "",
    iSIngredientDetailsActive: false,
};

export const reducer: Reducer<
    IIngredientDetailsState,
    IngredientDetailsActionsType
> = (state = initialState, action) => {
    switch (action.type) {
        case ActionTypes.SHOW_INGREDIENT_DETAILS: {
            return {
                ...state,
                iSIngredientDetailsActive: true,
                ingredientDetailsId: action.payload,
            };
        }
        case ActionTypes.HIDE_INGREDIENT_DETAILS: {
            return {
                ...state,
                iSIngredientDetailsActive: false,
                ingredientDetailsId: "",
            };
        }
        default:
            return state;
    }
};
