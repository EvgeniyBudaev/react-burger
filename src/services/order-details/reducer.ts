import { Reducer } from "redux";
import { ActionTypes, OrderDetailsActionsType } from "services/order-details";
import { IOrderDetails } from "types/order";
import { IError } from "types/error";

export interface IOrderDetailsState {
    details: IOrderDetails;
    detailsRequest: boolean;
    detailsFailed: boolean;
    detailsError: IError | null;
}

const initialState: IOrderDetailsState = {
    details: {
        name: "",
        order: {
            createdAt: "",
            _id: "",
            ingredients: [],
            name: "",
            number: undefined,
            owner: {
                createdAt: "",
                email: "",
                name: "",
                updatedAt: "",
            },
            status: undefined,
            updatedAt: "",
        },
        success: false,
    },
    detailsRequest: false,
    detailsFailed: false,
    detailsError: null,
};

export const reducer: Reducer<IOrderDetailsState, OrderDetailsActionsType> = (
    state = initialState,
    action
) => {
    switch (action.type) {
        case ActionTypes.GET_ORDER_DETAILS_REQUEST: {
            return {
                ...state,
                detailsRequest: true,
                detailsFailed: false,
                detailsError: null,
            };
        }
        case ActionTypes.GET_ORDER_DETAILS_SUCCESS:
            return {
                ...state,
                details: action.payload,
                detailsRequest: false,
            };
        case ActionTypes.GET_ORDER_DETAILS_FAILED:
            return {
                ...state,
                detailsRequest: false,
                detailsFailed: true,
                detailsError: action.payload,
            };
        default:
            return state;
    }
};
