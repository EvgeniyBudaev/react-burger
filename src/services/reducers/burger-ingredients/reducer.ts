import { AxiosError } from "axios";
import { Reducer } from "redux";
import { IIngredient } from "types/ingredient";
import {
    ActionTypes,
    BurgerIngredientsActionsType,
} from "services/reducers/burger-ingredients";

interface IBurgerIngredientsState {
    ingredients: IIngredient[];
    ingredientsRequest: boolean;
    ingredientsFailed: boolean;
    ingredientsError: AxiosError | null;
}

const initialState: IBurgerIngredientsState = {
    ingredients: [],
    ingredientsRequest: false,
    ingredientsFailed: false,
    ingredientsError: null,
};

export const reducer: Reducer<
    IBurgerIngredientsState,
    BurgerIngredientsActionsType
> = (state = initialState, action) => {
    switch (action.type) {
        case ActionTypes.GET_BURGER_INGREDIENTS_REQUEST: {
            return {
                ...state,
                ingredientsRequest: true,
                ingredientsFailed: false,
                ingredientsError: null,
            };
        }
        case ActionTypes.GET_BURGER_INGREDIENTS_SUCCESS:
            return {
                ...state,
                ingredients: action.payload,
                ingredientsRequest: false,
            };
        case ActionTypes.GET_BURGER_INGREDIENTS_FAILED:
            return {
                ...state,
                ingredientsRequest: false,
                ingredientsFailed: true,
                ingredientsError: action.payload,
            };
        default:
            return state;
    }
};
