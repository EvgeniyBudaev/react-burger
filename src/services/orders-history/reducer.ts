import { Reducer } from "redux";
import { ActionTypes, TOrdersHistoryActions } from "services/orders-history";
import { TFeed } from "types/feed";

interface IOrdersHistoryState {
    wsOrdersHistoryRequest: boolean;
    wsOrdersHistoryConnected: boolean;
    ordersHistory: TFeed | null;
    ordersHistoryRequest: boolean;
    ordersHistoryFailed: boolean;
    error: Event | null;
}

const initialState: IOrdersHistoryState = {
    wsOrdersHistoryRequest: false,
    wsOrdersHistoryConnected: false,
    ordersHistory: null,
    ordersHistoryRequest: false,
    ordersHistoryFailed: false,
    error: null,
};

export const reducer: Reducer<IOrdersHistoryState, TOrdersHistoryActions> = (
    state = initialState,
    action
) => {
    switch (action.type) {
        case ActionTypes.WS_ORDERS_HISTORY_CONNECTION_START: {
            return {
                ...state,
                wsOrdersHistoryRequest: true,
                error: null,
            };
        }
        case ActionTypes.WS_ORDERS_HISTORY_CONNECTION_STOP: {
            return {
                ...state,
                wsOrdersHistoryConnected: false,
            };
        }
        case ActionTypes.WS_ORDERS_HISTORY_CONNECTION_SUCCESS: {
            return {
                ...state,
                wsOrdersHistoryConnected: true,
                error: null,
            };
        }
        case ActionTypes.WS_ORDERS_HISTORY_CONNECTION_ERROR: {
            return {
                ...state,
                wsOrdersHistoryRequest: false,
                error: action.payload,
            };
        }
        case ActionTypes.WS_ORDERS_HISTORY_CONNECTION_CLOSED: {
            return {
                ...state,
                wsOrdersHistoryConnected: false,
                error: null,
            };
        }
        case ActionTypes.WS_GET_ORDERS_HISTORY: {
            return {
                ...state,
                wsOrdersHistoryRequest: false,
                ordersHistory: {
                    ...state.ordersHistory,
                    orders: action.payload.orders,
                    success: true,
                    total: action.payload.total,
                    totalToday: action.payload.totalToday,
                },
            };
        }
        default:
            return state;
    }
};
