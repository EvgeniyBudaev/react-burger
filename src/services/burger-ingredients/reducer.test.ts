import {
    ActionTypes,
    IBurgerIngredientsState,
} from "services/burger-ingredients";
import { reducer } from "./reducer";

const initialState: IBurgerIngredientsState = {
    ingredients: [],
    ingredientsRequest: false,
    ingredientsFailed: false,
    ingredientsError: null,
};

describe("burger-ingredients reducer", () => {
    it("should return the initial state", () => {
        expect(initialState).toMatchSnapshot();
    });

    it("should handle GET_BURGER_INGREDIENTS_REQUEST", () => {
        expect(
            reducer(initialState, {
                type: ActionTypes.GET_BURGER_INGREDIENTS_REQUEST,
            })
        ).toEqual(
            expect.objectContaining({
                ...initialState,
                ingredientsRequest: true,
                ingredientsFailed: false,
                ingredientsError: null,
            })
        );
    });

    it("should handle GET_BURGER_INGREDIENTS_SUCCESS", () => {
        expect(
            reducer(initialState, {
                type: ActionTypes.GET_BURGER_INGREDIENTS_SUCCESS,
                payload: [
                    {
                        calories: 643,
                        carbohydrates: 85,
                        count: 0,
                        _id: "60d3b41abdacab0026a733c8",
                        id: "63d3aca1-7c2d-4f98-a932-7fd94ce4a850",
                        image: "https://code.s3.yandex.net/react/code/meat-03.png",
                        image_large:
                            "https://code.s3.yandex.net/react/code/meat-03-large.png",
                        image_mobile:
                            "https://code.s3.yandex.net/react/code/meat-03-mobile.png",
                        fat: 26,
                        name: "Филе Люминесцентного тетраодонтимформа",
                        price: 988,
                        proteins: 44,
                        type: "main",
                        __v: 0,
                    },
                ],
            })
        ).toEqual(
            expect.objectContaining({
                ...initialState,
                ingredientsRequest: false,
                ingredients: [
                    {
                        calories: 643,
                        carbohydrates: 85,
                        count: 0,
                        _id: "60d3b41abdacab0026a733c8",
                        id: "63d3aca1-7c2d-4f98-a932-7fd94ce4a850",
                        image: "https://code.s3.yandex.net/react/code/meat-03.png",
                        image_large:
                            "https://code.s3.yandex.net/react/code/meat-03-large.png",
                        image_mobile:
                            "https://code.s3.yandex.net/react/code/meat-03-mobile.png",
                        fat: 26,
                        name: "Филе Люминесцентного тетраодонтимформа",
                        price: 988,
                        proteins: 44,
                        type: "main",
                        __v: 0,
                    },
                ],
            })
        );
    });

    it("should handle GET_BURGER_INGREDIENTS_FAILED", () => {
        expect(
            reducer(initialState, {
                type: ActionTypes.GET_BURGER_INGREDIENTS_FAILED,
                payload: {
                    success: false,
                    error: {
                        body: "error body",
                        message: "error message",
                    },
                },
            })
        ).toEqual(
            expect.objectContaining({
                ...initialState,
                ingredientsRequest: false,
                ingredientsFailed: true,
                ingredientsError: {
                    success: false,
                    error: {
                        body: "error body",
                        message: "error message",
                    },
                },
            })
        );
    });
});
