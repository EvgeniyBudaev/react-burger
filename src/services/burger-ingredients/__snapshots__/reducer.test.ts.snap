// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`burger-ingredients reducer should return the initial state 1`] = `
Object {
  "ingredients": Array [],
  "ingredientsError": null,
  "ingredientsFailed": false,
  "ingredientsRequest": false,
}
`;
